src/sources/aws_kinesis_firehose/filters.rs

```diff
@@ -0,0 +1,157 @@
+use super::{
+    errors::{Parse, RequestError},
+    handlers,
+    models::{FirehoseRequest, FirehoseResponse},
+};
+use crate::{
+    internal_events::{AwsKinesisFirehoseRequestError, AwsKinesisFirehoseRequestReceived},
+    Pipeline,
+};
+use bytes::{buf::BufExt, Bytes};
+use chrono::Utc;
+use flate2::read::GzDecoder;
+use snafu::ResultExt;
+use std::convert::Infallible;
+use std::io;
+use warp::http::StatusCode;
+use warp::Filter;
+
+/// Handles routing of incoming HTTP requests from AWS Kinesis Firehose
+pub fn firehose(
+    access_key: Option<String>,
+    out: Pipeline,
+) -> impl Filter<Extract = impl warp::Reply, Error = Infallible> + Clone {
+    warp::post()
+        .and(emit_received())
+        .and(authenticate(access_key))
+        .and(warp::header("X-Amz-Firehose-Request-Id"))
+        .and(warp::header("X-Amz-Firehose-Source-Arn"))
+        .and(
+            warp::header("X-Amz-Firehose-Protocol-Version")
+                .and_then(|version: String| async move {
+                    match version.as_str() {
```

Comment:
This didn't quite work as it gives me a `&String` when I want to compare it to a `&str`. Is there something I'm missing?

Reply:
This not works because the compiler can not apply [Deref](https://doc.rust-lang.org/std/string/struct.String.html#deref) in `match`. So [String::as_str](https://doc.rust-lang.org/std/string/struct.String.html#method.as_str) need to be used or slice with reference `&version[..]`, but `as_str` looks better :slightly_smiling_face:

Reply:
> This not works because the compiler can not apply [Deref](https://doc.rust-lang.org/std/string/struct.String.html#deref) in `match`

Thanks @fanatid, do you have a reference that mentions Deref not working in `match`? I didn't expect that not to work, so that's an interesting titbit to learn.

Reply:
I do not have reference where `match` mentioned with `Deref`, it's my assumption from the documentation on [Deref](https://doc.rust-lang.org/std/ops/trait.Deref.html) page.

> If T implements Deref<Target = U>, and x is a value of type T, then:
>    Values of type &T are coerced to values of type &U

From this, if the compiler can calculate output type `Deref` will work, example: `let str_type: &str = &string_type;` or when we call function like `bar(&string_type)` with signature `bar(_: &str)` -- in both cases output type can be calculated. But looks like type calculation not works for `match` in our case (I guess because we have `_` in arm).